spread/openstack.go

```diff
@@ -0,0 +1,631 @@
+  - test -d /etc/ssh/sshd_config.d && echo 'PermitRootLogin=yes' > /etc/ssh/sshd_config.d/00-spread.conf
+  - test -d /etc/ssh/sshd_config.d && echo 'PasswordAuthentication=yes' >> /etc/ssh/sshd_config.d/00-spread.conf
+  - pkill -o -HUP sshd || true
+`
+const openstackReadyMarker = "MACHINE-IS-READY"
+const openstackNameLayout = "Jan021504.000000"
+const openstackDefaultFlavor = "m1.medium"
+
+type openstackImage struct {
+	Project string
+	Name    string
+	Family  string
+	Terms   []string
+}
+
+type openstackImagesCache struct {
+	mu     sync.Mutex
+	ready  bool
+	images []openstackImage
+	err    error
+}
+
+func openstackName() string {
+	return strings.ToLower(strings.Replace(time.Now().UTC().Format(openstackNameLayout), ".", "-", 1))
+}
+
+func findErrorTitle(node *html.Node) string {
+	if node.Type == html.ElementNode && node.Data == "title" {
+		return node.FirstChild.Data
+	}
+
+	for child := node.FirstChild; child != nil; child = child.NextSibling {
+		title := findErrorTitle(child)
+		if title != "" {
+			return title
+		}
+	}
+
+	return ""
+}
+
+// error messages returned by openstack api are html
+// which contain title and details related to the error
+func errorTitle(msg string) string {
+	node, err := html.Parse(strings.NewReader(msg))
+	if err != nil {
+		return ""
+	}
+	return findErrorTitle(node)
+}
+
+func (p *openstackProvider) findFlavor(flavorName string) (nova.Entity, error) {
+	var flavor nova.Entity
+
+	flavors, err := p.computeClient.ListFlavors()
+	if err != nil {
+		return flavor, fmt.Errorf("failed to retrieve flavors list: %s", errorTitle(err.Error()))
+	}
+
+	for _, f := range flavors {
+		if f.Name == flavorName {
+			flavor = f
+			break
+		}
+	}
+
+	if flavor.Id == "" {
+		return flavor, fmt.Errorf("specified flavor not found: %s", flavorName)
+	}
+
+	return flavor, nil
+}
+
+func (p *openstackProvider) findNetwork(name string) (neutron.NetworkV2, error) {
+	var serverNetwork neutron.NetworkV2
+
+	networks, err := p.networkClient.ListNetworksV2()
+	if err != nil {
+		return serverNetwork, fmt.Errorf("failed to retrieve networks list: %s", errorTitle(err.Error()))
+	}
+
+	// When there are not networks defined, the first network which is not external
+	// is returned (external networks could not be allowed to request)
+	for _, net := range networks {
+		if (name == "" && net.External == false) || (name != "" && name == net.Name) {
+			return net, nil
+		}
+	}
+	return serverNetwork, &FatalError{fmt.Errorf("no valid network found to create floating IP")}
+}
+
+func (p *openstackProvider) findImage(imageName string) (glance.ImageDetail, error) {
+	var sameImage glance.ImageDetail
+	var lastImage glance.ImageDetail
+	var lastCreatedDate time.Time
+
+	images, err := p.imageClient.ListImagesDetail()
+	if err != nil {
+		return sameImage, fmt.Errorf("failed to retrieve images list: %s", errorTitle(err.Error()))
+	}
+
+	for _, i := range images {
```

Comment:
I wonder if this code should be slightly more elaborate and follow `googleProvider:image` or `linode:tempate()`. linode is simpler and just does a prefix search but afaict all do more than just check for "contains"

Reply:
@mvo5 could you please elaborate a bit more this? In openstack the images dont have family or project associated as in gce, so because of that I used the contain.

Reply:
Sorry, I was mostly wondering what contraints there are about image names, I created a unit test for the code now so that we can explore various test cases and examples :)